File: internal/sentence/command.go
package sentence

import (
	"regexp"
	"strings"
)

// The &-command layer. segman (the segmenter) decides *boundaries* — a block
// command is its own segment; an inline command rides inside its host
// sentence. This file owns *meaning*: recognizing a stored sentence as a
// command and parsing its fields (keyword, optional #slug, {...} args).
//
// Grammar (see TEX_COMMANDS_PLAN.md and PLACEHOLDER_PLAN.md):
//
//	&title{text}
//	&part#slug{text}{label?}
//	&chapter#slug{text}{label?}
//	&anchor#slug{label?}{details?}                       // details reserved, unrendered
//	&reference#slug{notes?}                              // inline only
//	&placeholder#slug{unit}{size?}{label?}{details?}     // see ParsePlaceholder
//	&snippet#id{label}                                   // canon region opener (VARIATIONS_PLAN.md)
//	&end#slug                                            // region terminator (SCRATCHPAD_PLAN.md)
//
// Argument vocabulary: {text} renders in the book, {label} shows in the
// outline, {details} is auxiliary metadata (a placeholder's details overlay;
// an anchor's details are parsed but unrendered for now).
//
// A '&' begins a command only when immediately followed by an exact keyword
// and then '#' or '{'. #slug is [a-z0-9-]+. Block vs inline is a segmentation
// concern (handled in segman); here we only parse.

// CommandKind identifies a recognized command keyword.
type CommandKind string

const (
	CmdTitle       CommandKind = "title"
	CmdPart        CommandKind = "part"
	CmdChapter     CommandKind = "chapter"
	CmdAnchor      CommandKind = "anchor"
	CmdReference   CommandKind = "reference"
	CmdMeta        CommandKind = "meta"
	CmdPlaceholder CommandKind = "placeholder"
	CmdSnippet     CommandKind = "snippet"
	CmdEnd         CommandKind = "end"
)

// blockCommandKinds are the commands that stand alone as their own sentence
// when they occupy their own line. (anchor is block only when alone on its
// line — that distinction is made by segman; a stored block-anchor sentence
// is recognized here like any other block command.) &meta is block but
// renders as nothing — it carries a book-wide setting.
var blockCommandKinds = map[CommandKind]bool{
	CmdTitle:       true,
	CmdPart:        true,
	CmdChapter:     true,
	CmdAnchor:      true,
	CmdMeta:        true,
	CmdPlaceholder: true, // block iff sole line content, same as anchor (segman's call)
	CmdSnippet:     true, // block iff sole line content; canon region opener (VARIATIONS_PLAN.md)
	CmdEnd:         true, // block iff sole line content; invisible region terminator
}

// Command is a parsed &-command. Slug is the author's static #slug ("" if the
// author gave none — callers derive an auto-slug from Args in that case).
// Args are the brace groups in order: e.g. chapter → [label, desc?].
type Command struct {
	Kind CommandKind
	Slug string   // "" when no #slug was written
	Args []string // brace-group contents, in source order
	Raw  string   // the exact matched command token
}

var (
	// slugPattern: a static slug is lowercase letters, digits, and dashes.
	slugPattern = regexp.MustCompile(`^[a-z0-9-]+$`)
	// commandNames matched at the start of a token.
	commandNames = []CommandKind{CmdTitle, CmdPart, CmdChapter, CmdAnchor, CmdReference, CmdMeta, CmdPlaceholder, CmdSnippet, CmdEnd}
)

// PlaceholderSpec is the interpreted argument list of a &placeholder command:
// {unit}{size?}{label?}{details?}. The size arg is positional but detected by
// value: an arg that exactly matches a size keyword is the size, anything
// else is the label (a label that IS literally a size keyword can force the
// default by writing the size explicitly). Valid=false means the command is
// mis-syntaxed and renders as literal prose.
type PlaceholderSpec struct {
	Unit    string // "sentences" or "paragraphs"
	Size    string // canonical size keyword; "m" when omitted
	Count   int    // resolved count of Unit
	Label   string // outline text (placeholder lists exactly like an anchor)
	Details string // overlaid on the rendered placeholder region
	Valid   bool
}

// Placeholder t-shirt sizes. Sentences double-ish; paragraphs are Fibonacci.
// The asymmetry is deliberate (PLACEHOLDER_PLAN.md).
var placeholderSentenceCounts = map[string]int{
	"xs": 1, "s": 2, "m": 3, "l": 5, "xl": 10, "xxl": 20, "xxxl": 40,
}
var placeholderParagraphCounts = map[string]int{
	"xs": 1, "s": 2, "m": 3, "l": 5, "xl": 8, "xxl": 13, "xxxl": 21,
}

// ParsePlaceholder interprets a parsed &placeholder command's args. Mirrors
// web/js/command.js placeholderSpec — keep in lockstep.
func ParsePlaceholder(cmd Command) PlaceholderSpec {
	spec := PlaceholderSpec{Size: "m"}
	if cmd.Kind != CmdPlaceholder || len(cmd.Args) == 0 || len(cmd.Args) > 4 {
		return spec
	}
	spec.Unit = strings.TrimSpace(cmd.Args[0])
	counts, ok := map[string]map[string]int{
		"sentences":  placeholderSentenceCounts,
		"paragraphs": placeholderParagraphCounts,
	}[spec.Unit]
	if !ok {
		return spec
	}
	rest := cmd.Args[1:]
	if len(rest) > 0 {
		if _, isSize := counts[strings.TrimSpace(rest[0])]; isSize {
			spec.Size = strings.TrimSpace(rest[0])
			rest = rest[1:]
		}
	}
	if len(rest) > 0 {
		spec.Label = rest[0]
		rest = rest[1:]
	}
	if len(rest) > 0 {
		spec.Details = rest[0]
		rest = rest[1:]
	}
	if len(rest) > 0 {
		// More args than {unit}{size}{label}{details} can absorb.
		return PlaceholderSpec{Size: "m"}
	}
	spec.Count = counts[spec.Size]
	spec.Valid = true
	return spec
}

// IsBlockCommandText reports whether a stored sentence is a block &-command
// (its whole text is one command token). Used to route block-command
// sentences through the own-line, blank-line-gapped tokenize/reconstruct path.
func IsBlockCommandText(text string) bool {
	cmd, ok := ParseCommand(strings.TrimSpace(text))
	if !ok {
		return false
	}
	// A block sentence's text is exactly the command (nothing trailing). An
	// inline anchor sharing a line never reaches here as its own sentence.
	return blockCommandKinds[cmd.Kind] && cmd.Raw == strings.TrimSpace(text)
}

// StaticSlug is a slug the author wrote (#slug) on a block command, paired
// with the sentence that carries it and the command kind. Extracted at
// migration time into the slug index (TEX_COMMANDS_PLAN.md §3).
type StaticSlug struct {
	Slug       string
	SentenceID string
	Kind       CommandKind
}

// Reference is an inline &reference#slug{notes} occurrence: the target slug,
// the notes text, and the char range within its host sentence's text.
type Reference struct {
	Slug       string
	Notes      string
	Start, End int // rune-index range in the host sentence text
}

// FindInlineCommands scans a sentence's text for inline &reference and
// &anchor commands (those that are NOT the whole sentence — inline ones ride
// inside prose). Returns each with its rune range. A block sentence (the whole
// text is one command) yields nothing here; block commands are handled
// separately (slug index / outline).
func FindInlineCommands(text string) (refs []Reference, anchors []Reference) {
	runes := []rune(text)
	// If the entire trimmed text is one block command, it's not inline.
	if IsBlockCommandText(text) {
		return nil, nil
	}
	for i := 0; i < len(runes); i++ {
		if runes[i] != '&' {
			continue
		}
		cmd, ok := ParseCommand(string(runes[i:]))
		if !ok {
			continue
		}
		end := i + len([]rune(cmd.Raw))
		notes := ""
		if len(cmd.Args) > 0 {
			notes = cmd.Args[0]
		}
		switch cmd.Kind {
		case CmdReference:
			refs = append(refs, Reference{Slug: cmd.Slug, Notes: notes, Start: i, End: end})
		case CmdAnchor:
			anchors = append(anchors, Reference{Slug: cmd.Slug, Notes: notes, Start: i, End: end})
		}
		i = end - 1 // skip past the matched command
	}
	return refs, anchors
}

// FindReferences returns just the reference slugs across a set of sentences,
// for migration-time dangling detection. Ignores anchors.
func FindReferences(ids []string, textByID map[string]string) []Reference {
	var out []Reference
	for _, id := range ids {
		text, ok := textByID[id]
		if !ok {
			continue
		}
		refs, _ := FindInlineCommands(text)
		out = append(out, refs...)
	}
	return out
}

// ExtractStaticSlugs walks (sentenceID -> text) block-command sentences and
// returns each author-written static #slug with its sentence and kind. Only
// static slugs are returned — a command with no #slug contributes nothing
// (its auto-slug is computed on read, never stored). Invalid slugs are
// skipped (they'd have failed validation upstream). Order follows the
// provided ids slice so results are deterministic.
func ExtractStaticSlugs(ids []string, textByID map[string]string) []StaticSlug {
	var out []StaticSlug
	for _, id := range ids {
		text, ok := textByID[id]
		if !ok {
			continue
		}
		cmd, ok := ParseCommand(strings.TrimSpace(text))
		if !ok || !blockCommandKinds[cmd.Kind] || cmd.Raw != strings.TrimSpace(text) {
			continue
		}
		if cmd.Slug == "" || !ValidSlug(cmd.Slug) {
			continue
		}
		out = append(out, StaticSlug{Slug: cmd.Slug, SentenceID: id, Kind: cmd.Kind})
	}
	return out
}

// ParseCommand parses a single command token at the START of s. It returns the
// command and true if s begins with a recognized command; trailing content
// after the command is ignored by the parse (cmd.Raw delimits what matched).
// Returns ok=false for a literal '&' (no keyword, or no '#'/'{' delimiter).
func ParseCommand(s string) (Command, bool) {
	runes := []rune(s)
	if len(runes) == 0 || runes[0] != '&' {
		return Command{}, false
	}
	kind, after := matchKeyword(runes)
	if kind == "" {
		return Command{}, false
	}

	i := after
	var slug string
	if i < len(runes) && runes[i] == '#' {
		i++
		start := i
		if kind == CmdEnd {
			// 'end' may be a bare #slug token with no {...} groups, so its
			// slug self-terminates on the slug charset [a-z0-9-].
			for i < len(runes) && isSlugRune(runes[i]) {
				i++
			}
		} else {
			for i < len(runes) && runes[i] != '{' {
				i++
			}
		}
		slug = string(runes[start:i])
	}

	// One or more {...} groups.
	var args []string
	for i < len(runes) && runes[i] == '{' {
		depth := 0
		start := i + 1
		for i < len(runes) {
			switch runes[i] {
			case '{':
				depth++
			case '}':
				depth--
				if depth == 0 {
					args = append(args, string(runes[start:i]))
					i++
					goto nextGroup
				}
			}
			i++
		}
		// Unterminated group.
		return Command{}, false
	nextGroup:
	}

	if len(args) == 0 && !(kind == CmdEnd && slug != "") {
		return Command{}, false
	}
	return Command{
		Kind: kind,
		Slug: slug,
		Args: args,
		Raw:  string(runes[:i]),
	}, true
}

// matchKeyword returns the command keyword at the start of runes (after '&')
// and the index just past it, requiring the next char to be '#' or '{'.
func matchKeyword(runes []rune) (CommandKind, int) {
	for _, name := range commandNames {
		end := 1 + len([]rune(string(name)))
		if end >= len(runes) {
			continue
		}
		if string(runes[1:end]) != string(name) {
			continue
		}
		if runes[end] == '#' || runes[end] == '{' {
			return name, end
		}
	}
	return "", 0
}

// ValidSlug reports whether a #slug matches the required charset [a-z0-9-]+.
func ValidSlug(slug string) bool {
	return slugPattern.MatchString(slug)
}

// isSlugRune reports whether r is in the #slug charset [a-z0-9-].
func isSlugRune(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '-'
}

File: web/js/command.js
/**
 * &-command parsing (frontend mirror of internal/sentence/command.go).
 *
 * segman decides boundaries; this parses a stored sentence into its command
 * fields for rendering. A '&' begins a command only when immediately followed
 * by an exact keyword and then '#' or '{' — so "Smith & Sons", "R&D", and
 * "A &chapter of accidents" are literal prose, not commands.
 *
 * Grammar (see TEX_COMMANDS_PLAN.md and PLACEHOLDER_PLAN.md):
 *   &title{text}
 *   &part#slug{text}{label?}
 *   &chapter#slug{text}{label?}
 *   &anchor#slug{label?}{details?}                    // details reserved, unrendered
 *   &reference#slug{notes?}                           // inline only
 *   &placeholder#slug{unit}{size?}{label?}{details?}  // see placeholderSpec
 *   &end#slug                                         // region terminator (SCRATCHPAD_PLAN.md)
 *
 * Argument vocabulary: {text} renders in the book, {label} shows in the
 * outline, {details} is auxiliary metadata (a placeholder's details overlay;
 * an anchor's details are parsed but unrendered for now).
 */

const WriteSysCommand = {
  // 'snippet' is the legacy spelling of 'sketch' (the snippet→sketch
  // rename) — parsed as the SAME command and valid input forever; parse()
  // normalizes the kind to 'sketch' so every consumer stays single-kind.
  KEYWORDS: ['title', 'part', 'chapter', 'anchor', 'reference', 'meta', 'placeholder', 'sketch', 'snippet', 'end'],
  // Block commands stand alone as their own sentence when on their own line.
  // &meta is block but renders as nothing (it carries a setting). anchor,
  // placeholder, and end are block only when sole line content (segman).
  BLOCK: { title: true, part: true, chapter: true, anchor: true, meta: true, placeholder: true, sketch: true, snippet: true, end: true },

  // Placeholder t-shirt sizes. Sentences double-ish; paragraphs are
  // Fibonacci. The asymmetry is deliberate (PLACEHOLDER_PLAN.md).
  PLACEHOLDER_SENTENCES: { xs: 1, s: 2, m: 3, l: 5, xl: 10, xxl: 20, xxxl: 40 },
  PLACEHOLDER_PARAGRAPHS: { xs: 1, s: 2, m: 3, l: 5, xl: 8, xxl: 13, xxxl: 21 },

  SLUG_RE: /^[a-z0-9-]+$/,

  // parse a command token at the START of s. Returns
  // { kind, slug, args, raw } or null if s does not begin with a command.
  parse(s) {
    const chars = Array.from(s);
    if (chars.length === 0 || chars[0] !== '&') return null;

    let kind = null;
    let i = 0;
    for (const kw of this.KEYWORDS) {
      const end = 1 + kw.length;
      if (end >= chars.length) continue;
      if (chars.slice(1, end).join('') !== kw) continue;
      if (chars[end] === '#' || chars[end] === '{') { kind = kw; i = end; break; }
    }
    if (kind === null) return null;
    if (kind === 'snippet') kind = 'sketch'; // legacy spelling — one internal kind

    let slug = '';
    if (i < chars.length && chars[i] === '#') {
      i++;
      const start = i;
      if (kind === 'end') {
        // 'end' may be a bare #slug token with no {...} groups, so its slug
        // self-terminates on the slug charset [a-z0-9-].
        while (i < chars.length && /[a-z0-9-]/.test(chars[i])) i++;
      } else {
        while (i < chars.length && chars[i] !== '{') i++;
      }
      slug = chars.slice(start, i).join('');
    }

    const args = [];
    while (i < chars.length && chars[i] === '{') {
      let depth = 0;
      const start = i + 1;
      let closed = false;
      while (i < chars.length) {
        if (chars[i] === '{') depth++;
        else if (chars[i] === '}') {
          depth--;
          if (depth === 0) { args.push(chars.slice(start, i).join('')); i++; closed = true; break; }
        }
        i++;
      }
      if (!closed) return null; // unterminated group
    }
    if (args.length === 0 && !(kind === 'end' && slug)) return null;

    return { kind, slug, args, raw: chars.slice(0, i).join('') };
  },

  // isBlockCommandText: the whole sentence is one block command (nothing
  // trailing). Mirrors IsBlockCommandText in Go.
  isBlockCommandText(text) {
    const t = text.trim();
    const cmd = this.parse(t);
    return !!cmd && !!this.BLOCK[cmd.kind] && cmd.raw === t;
  },

  validSlug(slug) {
    return this.SLUG_RE.test(slug);
  },

  // placeholderSpec interprets a &placeholder arg list {unit}{size?}{label?}
  // {details?}. The size arg is positional but detected by value: an arg that
  // exactly matches a size keyword is the size, anything else is the label (a
  // label that IS literally a size keyword can force the default by writing
  // the size explicitly). valid=false means mis-syntax → render as literal
  // prose. Mirrors Go ParsePlaceholder — keep in lockstep.
  placeholderSpec(args) {
    const spec = { unit: '', size: 'm', count: 0, label: '', details: '', valid: false };
    if (!args || args.length === 0 || args.length > 4) return spec;
    spec.unit = String(args[0]).trim();
    const counts = spec.unit === 'sentences' ? this.PLACEHOLDER_SENTENCES
      : (spec.unit === 'paragraphs' ? this.PLACEHOLDER_PARAGRAPHS : null);
    if (!counts) return spec;
    let rest = args.slice(1);
    if (rest.length > 0 && Object.prototype.hasOwnProperty.call(counts, String(rest[0]).trim())) {
      spec.size = String(rest[0]).trim();
      rest = rest.slice(1);
    }
    if (rest.length > 0) { spec.label = rest[0]; rest = rest.slice(1); }
    if (rest.length > 0) { spec.details = rest[0]; rest = rest.slice(1); }
    if (rest.length > 0) return { unit: '', size: 'm', count: 0, label: '', details: '', valid: false };
    spec.count = counts[spec.size];
    spec.valid = true;
    return spec;
  },

  // segmentFragments splits a sentence's effective text into an ordered list
  // of render fragments (SUGGESTION_RENDER_PLAN.md). Each fragment is one of:
  //   { kind:'command', cmd }              a block command (title/part/chapter/
  //                                        anchor/meta) on its own line
  //   { kind:'prose', text, marker }       a prose run; marker is '' | '\n\n'
  //                                        | '\n\t' (its leading structural
  //                                        break, for the paragraph grouping)
  // Blocks are separated by \n\n or \n\t. A block whose whole content is a
  // block command becomes a command fragment; everything else is prose.
  segmentFragments(text) {
    const frags = [];
    const s = String(text == null ? '' : text);
    // Split keeping the delimiters so we know each block's leading marker.
    // Tokens alternate: [block, delim, block, delim, ...].
    const parts = s.split(/(\n\n|\n\t)/);
    let marker = '';
    for (let i = 0; i < parts.length; i++) {
      const piece = parts[i];
      if (piece === '\n\n' || piece === '\n\t') { marker = piece; continue; }
      if (piece === '') continue; // empty block (e.g. leading delimiter)
      const trimmed = piece.trim();
      const cmd = this.parse(trimmed);
      if (cmd && this.BLOCK[cmd.kind] && cmd.raw === trimmed) {
        frags.push({ kind: 'command', cmd, marker });
        marker = '';
        continue;
      }
      // Leading-command block form (canonicalize output): "&anchor{...}\nprose"
      // or "&sketch#x{...}\nprose" (sketch spelling included — parse()
      // normalizes it to kind 'sketch'). canonicalize joins a leading block
      // anchor/sketch to its paragraph with a single \n (canonicalize.js:45–47
      // / canonicalize.go — the same kinds, kept in lockstep), and segman
      // splits it into its own sentence. Mirror that split here so the render
      // preview matches the pushed/segmented result: emit the command as a
      // fragment, then the trailing prose as its paragraph.
      //
      // The block's leading marker (\n\n / \n\t) belongs to the PARAGRAPH, so it
      // rides on the PROSE fragment — that's what drives indent/section behavior
      // (e.g. a paragraph after a placeholder keeps its \n\t indent). The anchor
      // glyph is just a left-margin decoration on that paragraph and carries no
      // marker of its own.
      if (cmd && (cmd.kind === 'anchor' || cmd.kind === 'sketch')) {
        const after = trimmed.slice(cmd.raw.length);
        if (after.startsWith('\n')) {
          const prose = after.slice(1);
          frags.push({ kind: 'command', cmd, marker: '' });
          if (prose.trim() !== '') frags.push({ kind: 'prose', text: prose, marker });
          marker = '';
          continue;
        }
      }
      // Markdown # headers are deprecated and no longer rendered specially —
      // a '#' line is ordinary prose (convert to an &-command for a heading).
      // A block whose text BEGINS with a tab is explicitly an indented
      // paragraph, whatever delimiter preceded it — "\n\n\t" is a blank line
      // for raw-text readability (typically around a command line) plus a
      // tabbed paragraph, not a section break.
      frags.push({ kind: 'prose', text: piece, marker: piece.startsWith('\t') ? '\n\t' : marker });
      marker = '';
    }
    return frags;
  },

  // metaProperties: the fixed &meta vocabulary (mirror of Go's
  // metaProperties). A property mapped to an array of allowed values is
  // validated against it; a property mapped to null accepts any non-empty
  // value (e.g. a font name).
  META_PROPERTIES: {
    'chapter-align': ['left', 'center'],
    'part-align': ['left', 'center'],
    'title-align': ['left', 'center'],
    'divider-folios': ['on', 'off'],
    'font': null,
  },

  // extractSettings walks (id -> effectiveText) sentences, reads &meta
  // commands, and returns the validated {property: value} map. Last-wins for
  // a repeated property; unknown properties and out-of-range values dropped.
  // "Effective text" means the suggestion if one exists, else committed — so
  // a suggested &meta applies live and a suggested removal drops it.
  //
  // A sentence's effective text may be multi-block (a suggestion like
  // "&meta{...}\n\n&title{...}"), so we scan each \n\n / \n\t-separated block
  // for a whole-block &meta, not just whole-sentence meta.
  extractSettings(ids, effectiveTextById) {
    const values = {};
    const consume = (blockText) => {
      const t = String(blockText).trim();
      if (!t.startsWith('&meta')) return;
      const cmd = this.parse(t);
      if (!cmd || cmd.kind !== 'meta' || cmd.raw !== t) return;
      if (cmd.args.length < 2) return;
      const prop = (cmd.args[0] || '').trim();
      const val = (cmd.args[1] || '').trim();
      if (!(prop in this.META_PROPERTIES)) return;
      const allowed = this.META_PROPERTIES[prop];
      if (allowed && !allowed.includes(val)) return;
      if (!val) return;
      values[prop] = val; // last-wins
    };
    for (const id of ids) {
      const text = effectiveTextById[id];
      if (text == null) continue;
      // Split into blocks on \n\n or \n\t (the structural boundaries).
      String(text).split(/\n\n|\n\t/).forEach(consume);
    }
    return values;
  },

  // findInline scans a string for inline &reference / &anchor / &placeholder
  // commands (those not spanning the whole string) and returns [{kind, slug,
  // notes, args, raw, start, end}] in order. Used to render references as
  // links, anchors as markers, and placeholders as hatch regions within a
  // sentence's visible text.
  findInline(text) {
    const chars = Array.from(text);
    const out = [];
    // If the whole (trimmed) text is one block command, nothing is inline.
    if (this.isBlockCommandText(text)) return out;
    let i = 0;
    while (i < chars.length) {
      if (chars[i] !== '&') { i++; continue; }
      const cmd = this.parse(chars.slice(i).join(''));
      if (!cmd) { i++; continue; }
      const end = i + Array.from(cmd.raw).length;
      if (cmd.kind === 'reference' || cmd.kind === 'anchor' || cmd.kind === 'placeholder' || cmd.kind === 'sketch' || cmd.kind === 'end') {
        out.push({ kind: cmd.kind, slug: cmd.slug, notes: cmd.args[0] || '', args: cmd.args, raw: cmd.raw, start: i, end });
      }
      i = end;
    }
    return out;
  },

  // structuralForm describes how a block &-command renders as a heading-like
  // element, or null if it's ordinary content. Shape: { tag, cls, visible,
  // description }. The on-page heading shows ONLY the label — the description
  // is outline metadata (never rendered in the book).
  //   &title{name}              -> { h1, 'cmd-title',   name }
  //   &part#s{label}{desc?}     -> { h2, 'cmd-part',    label }
  //   &chapter#s{label}{desc?}  -> { h3, 'cmd-chapter', label }
  //   &anchor#s{desc?}          -> { div,'cmd-anchor',  desc }   (no label)
  // (Markdown # headers are deprecated — a '#' sentence is ordinary content.)
  structuralForm(text) {
    const t = String(text);
    const cmd = this.parse(t.trim());
    if (!cmd || !this.BLOCK[cmd.kind] || cmd.raw !== t.trim()) return null;
    switch (cmd.kind) {
      case 'title':
        return { tag: 'h1', cls: 'cmd-title', visible: cmd.args[0] || '', description: '' };
      case 'part':
        return { tag: 'h2', cls: 'cmd-part', visible: cmd.args[0] || '', description: cmd.args[1] || '' };
      case 'chapter':
        return { tag: 'h3', cls: 'cmd-chapter', visible: cmd.args[0] || '', description: cmd.args[1] || '' };
      case 'anchor':
        // A block anchor renders as a glyph (⚓), NOT its label text. The label
        // is outline metadata and shows on hover (title) + in the outline nav.
        return { tag: 'div', cls: 'cmd-anchor', visible: '', description: '', glyph: true, label: cmd.args[0] || '' };
      case 'sketch':
        // A canon-region opener (VARIATIONS_PLAN.md) renders exactly like a
        // block anchor — the extra class is a styling/selection hook.
        return { tag: 'div', cls: 'cmd-anchor cmd-sketch', visible: '', description: '', glyph: true, label: cmd.args[0] || '' };
      default:
        return null;
    }
  },
};

if (typeof window !== 'undefined') {
  window.WriteSysCommand = WriteSysCommand;
}
if (typeof module !== 'undefined' && module.exports) {
  module.exports = WriteSysCommand;
}
